plugins/utils/which-key.nix

```diff
@@ -55,257 +126,265 @@ with lib;
-          this is mostly relevant for keymaps that start with a native binding
-        '';
+    icons = {
+      breadcrumb = helpers.defaultNullOpts.mkStr "»" "Symbol used in the command line area that shows your active key combo";
+      separator = helpers.defaultNullOpts.mkStr "➜" "Symbol used between a key and it's label";
+      group = helpers.defaultNullOpts.mkStr "+" "Symbol prepended to a group";
+      ellipsis = helpers.defaultNullOpts.mkStr "…" "Symbol used for overflow ";
+      mappings = helpers.defaultNullOpts.mkBool true "Set to false to disable all mapping icons.";
+      rules =
+        helpers.defaultNullOpts.mkNullable (types.either (types.listOf types.attrs) types.bool) [ ]
+          "Icon rules. Set to false to disable all icons.";
+      colors = helpers.defaultNullOpts.mkBool true ''
+        Use the highlights from mini.icons.
+
+        When `false`, it will use `WhichKeyIcon` instead.
+      '';
+
+      keys = helpers.defaultNullOpts.mkNullable types.attrs {
+        Up = " ";
+        Down = " ";
+        Left = " ";
+        Right = " ";
+        C = "󰘴 ";
+        M = "󰘵 ";
+        D = "󰘳 ";
+        S = "󰘶 ";
+        CR = "󰌑 ";
+        Esc = "󱊷 ";
+        ScrollWheelDown = "󱕐 ";
+        ScrollWheelUp = "󱕑 ";
+        NL = "󰌑 ";
+        BS = "󰁮";
+        Space = "󱁐 ";
+        Tab = "󰌒 ";
+        F1 = "󱊫";
+        F2 = "󱊬";
+        F3 = "󱊭";
+        F4 = "󱊮";
+        F5 = "󱊯";
+        F6 = "󱊰";
+        F7 = "󱊱";
+        F8 = "󱊲";
+        F9 = "󱊳";
+        F10 = "󱊴";
+        F11 = "󱊵";
+        F12 = "󱊶";
+      } "Icons used by key format.";
+    };
+
+    show_help = helpers.defaultNullOpts.mkBool true "show a help message in the command line for using WhichKey";
+
+    show_keys = helpers.defaultNullOpts.mkBool true "show the currently pressed key and its label as a message in the command line";
 
     disable = {
-      buftypes = helpers.defaultNullOpts.mkListOf types.str [ ] "Disabled by default for Telescope";
-      filetypes = helpers.defaultNullOpts.mkListOf types.str [ ] "";
+      bt = helpers.defaultNullOpts.mkListOf types.str [ ] "Disabled by default for Telescope";
+      ft = helpers.defaultNullOpts.mkListOf types.str [ ] "";
     };
+
+    debug = helpers.defaultNullOpts.mkBool false "Enable wk.log in the current directory.";
   };
 
-  config =
+  # TODO: configure
+  settingsExample = { };
+
+  extraOptions = {
+    registrations = mkOption {
+      type = with types; attrsOf anything;
+      description = ''
+        This option is deprecated, use `settings.spec` instead.
+
+        Note: the keymap format has changed in v3 of which-key.
+      '';
+      visible = false;
+    };
+
+    window = mkOption {
+      type = with types; attrsOf anything;
+      description = ''
+        This option is deprecated, use `settings.win` instead.
+
+        Note: the window options have changed parameters in v3 of which-key.
+      '';
+      visible = false;
+    };
```

Comment:
The sheer quantity of options here is worrying to me (from a maintenance perspective). If any one of these changes upstream, we would have to figure out how to update it in our module - assuming we're even aware of the change!

While it's nice at a superficial level to have comprehensive type-checking and documentation in nixvim, it's my view that it isn't worth the drawbacks to do so.

I appreciate it would have taken some effort to check over them all, so thanks for your efforts regardless!

We're still having this debate internally, so take my opinion with a grain of salt.

Reply:
I guess my approach to covering the options was that when I got started with nix, nixvim helped me discover the options a plugin provided through the documentation and helped implementation a ton through the type definitions.

Reply:
Yes, historically we've taken the same approach of declaring options for pretty much everything.

It is nice to be able to reference nixvim's docs to discover plugin options however this comes with plenty of downsides too:

- no documentation is better than incorrect documentation
  - it's easy for us to get things wrong when manually adding options, or for our options to get out of date due to upstream changes
- type checking is nice, until it prevents defining valid config
  - having typed sub-options overrides the freeformType, somewhat defeating the point of rfc42
- sub-options are the _only_ reason we have to filter out null/empty values from the final `settings` value
  - this is a bigger issue though, as we'd either have to remove _all_ sub-options or find a way to only merge _defined_ options into the final value
  - the latter would require changes to nixpkgs `lib.evalModules` but is probably the correct approach
  - to clarify, all options end up in `config`; undefined options just `throw` when evaluated.
  - This isn't really relevant (the other two points are)

Reply:
@GaetanLepage @traxys Either of you have an opinion on the scope of options provided in the module? Originally went through and made sure all the plugin's options were defined explicitly, but there are drawbacks (as we've seen previously) to having it done.

Reply:
@MattSturgeon's explanation is quite exhaustive.
Actually, we (maintainers) haven't really settled on a clear decision on this topic.
It is quite hard to do so, as no solution is perfect. I also find a great value in having a lot of options defined. This comes very handy to the users.
However, I am more than aware of the maintenance burden that it implies.

Given the amount of options declared in nixvim today, I see no future where we shift entirely to empty `settings` submodules. In this sense, we will likely keep a hybrid approach, no matter what strategy we encourage/enforce.

In practice, my opinion is that while we agree on a clear way of doing this, plugin maintainers should be granted the freedom of doing what they want.
At the same time, you should be somehow responsible for trying to keep those options up to date.
All in all, if you want to keep those declarations, I am personally fine with it.

Reply:
> In practice, my opinion is that while we agree on a clear way of doing this, plugin maintainers should be granted the freedom of doing what they want.
> At the same time, you should be somehow responsible for trying to keep those options up to date.
> All in all, if you want to keep those declarations, I am personally fine with it.

I'm fine with this in the short term. It seems unfair to ask plugin maintainers to follow a policy we haven't fully decided on yet.

Reply:
> Sure you didn't mean _exhausting_? 😉

No haha ! I sincerely admire your surgically clear prose. It is always a joy to read !